README.md
REPLesent
=========

[![Build Status](https://travis-ci.org/marconilanna/REPLesent.svg)](https://travis-ci.org/marconilanna/REPLesent)
[![Codacy Badge](https://www.codacy.com/project/badge/2ec4acaf2e424baab3d0cb53d20d9df9)](https://www.codacy.com/public/marconilanna/REPLesent)

> **represent** *verb*
>  0. depict, portray, render, delineate, show, illustrate.
>  0. symbolize, stand for, personify, typify, embody.
>  0. point out, state, present, put forward.

REPLesent is a ~~neat little~~ powerful tool to build presentations using the Scala REPL.

Conceptualized and originally implemented while waiting almost two hours for my plane
to be de-iced, it was introduced during the Northeast Scala Symposium 2015 in Boston.

While clearly not a ~~Powerpoint~~ Keynote replacement, REPLesent is a good
option for training sessions and technical talks featuring live coding.
Its old-school looks were considered very cool by many conference attendants.

Features
--------

* Easy to write slides: a simple plain text file with minimal markup
* The full arsenal of navigation options: next, previous, first, last, jump to
* Builds (incremental slides)
* Slide number / total
* Text alignment: left, right, centered, flushed
* ANSI colors
* Horizontal rulers (thanks, [@daviscabral](https://github.com/daviscabral))
* Syntax highlighting
* Run code straight from slides directly in the REPL with a single keystroke.
No other presentation tool can do that for you!
* And the #1 requested feature: emoji! :-) I mean, :smile:

Quick Tour
----------

[![asciicast](https://asciinema.org/a/16690.png)](https://asciinema.org/a/16690)

Getting Started
---------------

REPLesent is distributed as a single `.scala` file and has no dependencies.

REPLesent was designed to be used only in conjunction with the Scala REPL.
It is not meant to be compiled as a standalone application
(`build.sbt` is only for running the unit tests).

We recommend you install and use the full Scala distribution
(the `scala` command) instead of just the `sbt` console.
Scala 2.11.4 or later and JDK 7 or later are the preferred versions. 
If you are using Windows, you need to run `scala` in
[ANSICON](https://github.com/adoxa/ansicon) or another console that
supports ANSI escape codes. REPLesent will not render correctly in
Windows' standard `cmd` console.

To get started, download and save to the same folder the files
`REPLesent.scala` and `REPLesent.txt`.
The file `emoji.txt` is optional, only needed to enable emoji support.

First, create an alias:

```sh
alias REPLesent='scala -Dscala.color -language:_ -nowarn -i REPLesent.scala'
```

Open the REPL and enter the two statements below:

```sh
$ REPLesent
Loading REPLesent.scala...
defined class REPLesent

Welcome to Scala version 2.11.5 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_31).
Type in expressions to have them evaluated.
Type :help for more information.

scala> val replesent = REPLesent(80, 25, intp=$intp)
replesent: REPLesent = REPLesent(80,25,REPLesent.txt)

scala> import replesent._
import replesent._
```

Do not forget to replace `80, 25` with the actual size of your terminal window.
The `intp=$intp` incantation is required to hook onto the REPL to run slide code.

Once all is done, type `f`, press `ENTER`, and follow the on-screen instructions.

**Experimental screen size auto-detection:**
For Unix-like systems, including Mac OS X and Linux, you can omit the screen size.
Does not work for Microsoft Windows systems.

Quick Reference Guide
---------------------

### Initialization options

Parameter | Type | Description | Default
--- | --- | --- | ---
`width` | `Int` | Terminal width in columns | `0` (Unix: autodetect, Windows: 80)
`height` | `Int` | Terminal height in rows | `0` (Unix: autodetect, Windows: 25)
`source` | `String` | The path to the presentation file | `"REPLesent.txt"`
`slideCounter` | `Boolean` | Whether to show the slide number | `false`
`slideTotal` | `Boolean` | Whether to show the total number of slides | `false`
`intp` | | A hook to the Scala REPL | No default, use magic value `$intp`

### Navigation commands

Command | Shortcut | Symbolic alias | Description
--- | --- | --- | ---
`next` | `n` | `>` | Go to next build/slide
`previous` | `p` | `<` | Go back to previous build/slide
`redraw` | `z` | | Redraw the current build/slide
`reload` | `y` | | Reload the slides from file
`Next` | `N` | `>>` | Go to next slide
`Previous` | `P` | `<<` | Go back to previous slide
`i next` | `i n` | | Advance i slides
`i previous` | `i p` | | Go back i slides
`i go` | `i g` | | Go to slide i
`first` | `f` | `|<` | Go to first slide
`last` | `l` | `>|` | Go to last slide
`Last` | `L` | `>>|` | Go to last build of last slide
`run` | `r` | `!!` | Execute code that appears on slide
`blank` | `b` | | Blank screen
`help` | `h` | `?` | This help message

### Separators and delimiters

Separator | Description
--- | ---
`---` | Separates slides
`--` | Separates builds
<code>```</code> | Delineates Scala code

### Text alignment

Command | Description
--- | ---
`<<` | Left-flushed text
`<` | Left-aligned text
`|` | Centered text
`>` | Right-aligned text
`>>` | Right-flushed text

A space separating the alignment command from the text is mandatory.

### ANSI colors

Escape code | Result
--- | ---
`\x` | Foreground color, where `x` is one of: `r`ed, `g`reen, `b`lue, `c`yan, `m`agenta, `y`ellow, blac`k`, `w`hite
`\X` | Background color, where capital `X` is one of the same as above
`\*` | Bold
`\_` | Underscore
`\!` | Reverse colors
`\s` | Resets to normal

### Horizontal rulers

Command | Description
--- | ---
`/` | A ruler across the slide length
`//` | A ruler across the entire screen width

An optional pattern may be specified immediately following the forward slash.
Unicode characters and ANSI color escapes (as above) are supported.

### Emoji

To enable emoji support, you will need a copy of the `emoji.txt` file.

Almost all shortcuts listed by the [Emoji cheat sheet](http://www.emoji-cheat-sheet.com/)
are supported.
You can look at `emoji.txt` for the definitive list of supported emoji.

Emojis can be combined with horizontal rulers and,
depending on your system fonts, ANSI colors, too.

Emoji has only be tested on Mac OS X Terminal.app and Gnome Terminal, YMMV.

### Multiple File Support

As your `REPLesent.txt` grows, you may find yourself needing to split slides into logical chunks for quick navigation.

In order to use this mode, simply place all your individual slide files into a directory (in this example, we'll be using the directory name `"slides"`), all ending with the file extension `.replesent`.

Then, initialize REPLesent with the additional parameter `source="slides"`, which will concatinate all files in ASCIIbetical order. Suggested practice is to name files with a leading sequence number, padding with zeros (01, 02, 03 ... 14, 15 ...)

One caveat is that since this is strict file concatination, slide separators will not be added automatically, so please remember to end your files with `---` or `--`.

### Vim Syntax Highlighting

Included in this repository is a `vim` directory, which contains a vim plugin for editing REPLesent slides.

This plugin depends on [derekwyatt/vim-scala](https://github.com/derekwyatt/vim-scala/) for Scala syntax highlighting.

Simply copy the files into your `~/.vim` directory to get started, or use [Vundle](https://github.com/VundleVim/Vundle.vim) (or similar) like:

    Plugin 'derekwyatt/vim-scala'
    Plugin 'marconilanna/REPLesent', {'rtp': 'vim/'}

Thanks
------

* [Davis Z. Cabral](https://github.com/daviscabral) for implementing
horizontal ruler support.
* [Noel Markham](https://github.com/noelmarkham) for the redraw command
* [Wojtek Pituła](https://github.com/Krever) for the reload command
* [Devon Stewart](https://github.com/blast-hardcheese/) for multi-file support, syntax highlighting plug-in for vim, and syntax highlighting enhancements.

License
-------

Copyright 2015-2017 Marconi Lanna

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# Change Log

## [Unreleased][unreleased]

## [1.2] - 2017-10-13
### Added
- `z` command: redraw the current build/slide (thanks @noelmarkham)
- `y` command: reload slide file from disk (thanks @Krever)
- A handful of improvements (thanks @blast-hardcheese)
    - Multi-file support
    - Syntax highlighting plug-in for vim
### Changed
- Syntax highlighting tweaks
- Update dependencies: Scala 2.12.3, Java 8, sbt 1.0.2, etc.

## [1.1] - 2015-02-25
### Added
- Emoji support

## [1.0] - 2015-02-19
### Added
- Updated tutorial slideshow
- Updated instructions in README.md
- Added Quick Reference Guide
- Added sample screencast tour
### Changed
- No space between a horizontal ruler command and its pattern

## [0.9] - 2015-02-18
### Added
- Support for running code straight from the slides

## [0.8] - 2015-02-17
### Added
- Horizontal ruler support (thanks @daviscabral)

## [0.7] - 2015-02-16
### Added
- Basic syntax highlighting

## [0.6] - 2015-02-10
### Added
- Slide counter/total

## [0.5] - 2015-02-09
### Added
- Line alignment

## [0.4] - 2015-02-09
### Added
- ANSI color support

## [0.3] - 2015-02-08
### Added
- Navigation options: by slides or by builds
- Unit tests

## [0.2] - 2015-02-01
### Added
- Experimental screen size auto-detection for Mac OS X, Linux, and Unix-like systems

## [0.1] - 2015-02-01
### Added
- Initial public release
- Basic navigation
- Build (incremental slides) support
- Margin (border) configuration
- Tutorial / sample slideshow

[unreleased]: https://github.com/marconilanna/REPLesent/compare/v1.2...HEAD
[1.2]: https://github.com/marconilanna/REPLesent/compare/v1.1...v1.2
[1.1]: https://github.com/marconilanna/REPLesent/compare/v1.0...v1.1
[1.0]: https://github.com/marconilanna/REPLesent/compare/v0.9...v1.0
[0.9]: https://github.com/marconilanna/REPLesent/compare/v0.8...v0.9
[0.8]: https://github.com/marconilanna/REPLesent/compare/v0.7...v0.8
[0.7]: https://github.com/marconilanna/REPLesent/compare/v0.6...v0.7
[0.6]: https://github.com/marconilanna/REPLesent/compare/v0.5...v0.6
[0.5]: https://github.com/marconilanna/REPLesent/compare/v0.4...v0.5
[0.4]: https://github.com/marconilanna/REPLesent/compare/v0.3...v0.4
[0.3]: https://github.com/marconilanna/REPLesent/compare/v0.2...v0.3
[0.2]: https://github.com/marconilanna/REPLesent/compare/v0.1...v0.2
[0.1]: https://github.com/marconilanna/REPLesent/tree/v0.1


REPLesent.scala
/*
 * Copyright 2015-2017 Marconi Lanna
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
case class REPLesent(
  width: Int = 0
, height: Int = 0
, source: String = "REPLesent.txt"
, slideCounter: Boolean = false
, slideTotal: Boolean = false
, intp: scala.tools.nsc.interpreter.IMain = null
) {
  import java.io.File
  import scala.util.matching.Regex
  import scala.util.{ Try, Success, Failure }

  private case class Config(
    top: String = "*"
  , bottom: String = "*"
  , sinistral: String = "* "
  , dextral: String = " *"
  , newline: String = System.lineSeparator
  , whiteSpace: String = " "
  , private val width: Int
  , private val height: Int
  ) {
    val (screenWidth, screenHeight): (Int, Int) = {
      val defaultWidth = 80
      val defaultHeight = 25

      if (width > 0 && height > 0) (width, height) else {
        // Experimental support for screen size auto-detection.
        // Supports only Unix-like systems, including Mac OS X and Linux.
        // Does not work with Microsoft Windows.
        val Array(h, w) = Try {
          import scala.sys.process._

          val stty = Seq("sh", "-c", "stty size < /dev/tty").!!

          stty.trim.split(' ') map (_.toInt)
        } getOrElse Array(0, 0)

        val screenWidth = Seq(width, w) find (_ > 0) getOrElse defaultWidth
        val screenHeight = Seq(height, h) find (_ > 0) getOrElse defaultHeight

        (screenWidth, screenHeight)
      }
    }

    private def fill(s: String): String = if (s.isEmpty) s else {
      val t = s * (screenWidth / s.length)
      t + s.take(screenWidth - t.length)
    }

    val topRow = fill(top) + newline
    val bottomRow = fill(bottom)

    val verticalSpace = screenHeight - 3 // accounts for header, footer, and REPL prompt
    val horizontalSpace = screenWidth - sinistral.length - dextral.length

    val blankLine = {
      val padding = if (dextral.isEmpty) "" else whiteSpace * horizontalSpace + dextral
      sinistral + padding + newline
    }
  }

  private val config = Config(width = width, height = height)

  private case class Line(content: String, length: Int, private val style: Line.Style) {
    override def toString: String = content
    def isEmpty: Boolean = content.isEmpty
    def render(margin: Int): String = style(this, margin)
  }

  private object Line {
    import scala.io.AnsiColor._

    protected sealed trait Style {
      import config.whiteSpace

      protected def horizontalSpace = config.horizontalSpace

      protected def fill(line: Line, left: Int, right: Int): String = {
        whiteSpace * left + line + whiteSpace * right
      }

      def apply(line: Line, margin: Int): String
    }

    private object HorizontalRuler extends Style {
      private val ansiBegin = RESET.head
      private val ansiEnd = RESET.last

      private val defaultPattern = Line("-")

      def apply(line: Line, margin: Int): String = {
        // Provides a default pattern if none was specified
        val pattern = if (line.isEmpty) defaultPattern else line

        val width = horizontalSpace - margin
        val repeats = width / pattern.length

        val content = pattern.toString * repeats

        var remaining = width - repeats * pattern.length
        var ansi = false
        var reset = ""

        val padding = pattern.toString takeWhile { c =>
          val continue = remaining > 0

          if (continue) c match {
            case `ansiEnd` if ansi => ansi = false
            case _ if ansi => // no-op
            case `ansiBegin` => ansi = true; reset = RESET
            case c if Character.isHighSurrogate(c) => // no-op
            case _ => remaining -= 1
          }

          continue
        }

        val left = margin / 2
        val right = margin - left

        val l = Line(content + padding + reset, width, LeftAligned)

        fill(l, left, right)
      }
    }

    private object FullScreenHorizontalRuler extends Style {
      def apply(line: Line, ignored: Int): String = HorizontalRuler(line, 0)
    }

    private object LeftFlushed extends Style {
      def apply(line: Line, ignored: Int): String = {
        val left = 0
        val right = horizontalSpace - line.length

        fill(line, left, right)
      }
    }

    private object LeftAligned extends Style {
      def apply(line: Line, margin: Int): String = {
        val left = margin / 2
        val right = horizontalSpace - left - line.length

        fill(line, left, right)
      }
    }

    private object Centered extends Style {
      def apply(line: Line, ignored: Int): String = {
        val margin = horizontalSpace - line.length

        val left = margin / 2
        val right = margin - left

        fill(line, left, right)
      }
    }

    private object RightAligned extends Style {
      def apply(line: Line, margin: Int): String = {
        val right = (margin + 1) / 2
        val left = horizontalSpace - right - line.length

        fill(line, left, right)
      }
    }

    private object RightFlushed extends Style {
      def apply(line: Line, ignored: Int): String = {
        val left = horizontalSpace - line.length
        val right = 0

        fill(line, left, right)
      }
    }

    private def style(line: String): (String, Style) = line match {
      case s if s startsWith "<< " => (s.drop(3), LeftFlushed)
      case s if s startsWith "< " => (s.drop(2), LeftAligned)
      case s if s startsWith "| " => (s.drop(2), Centered)
      case s if s startsWith "> " => (s.drop(2), RightAligned)
      case s if s startsWith ">> " => (s.drop(3), RightFlushed)
      case s if s startsWith "//" => (s.drop(2), FullScreenHorizontalRuler)
      case s if s startsWith "/" => (s.drop(1), HorizontalRuler)
      case s: String => (s, LeftAligned)
    }

    private val ansiEscape = """\\.""".r

    private val ansiColor = Map(
      'b' -> BLUE,
      'c' -> CYAN,
      'g' -> GREEN,
      'k' -> BLACK,
      'm' -> MAGENTA,
      'r' -> RED,
      'w' -> WHITE,
      'y' -> YELLOW,
      'B' -> BLUE_B,
      'C' -> CYAN_B,
      'G' -> GREEN_B,
      'K' -> BLACK_B,
      'M' -> MAGENTA_B,
      'R' -> RED_B,
      'W' -> WHITE_B,
      'Y' -> YELLOW_B,
      '!' -> REVERSED,
      '*' -> BOLD,
      '_' -> UNDERLINED
    )

    private def ansi(line: String): (String, Int) = {
      var drop = 0
      var reset = ""

      val content: String = ansiEscape.replaceAllIn(line, m =>
        m.matched(1) match {
          case c if ansiColor.contains(c) => drop += 2; reset = RESET; ansiColor(c)
          case 's' => drop += 2; RESET
          case '\\' => drop += 1; "\\\\"
          case c: Char => "\\\\" + c
        }
      )

      (content + reset, drop)
    }

    private val emojiEscape = """:([\w+\-]+):""".r

    private lazy val emojis: Map[String, String] = {
      Try {
        val emoji = io.Source.fromFile("emoji.txt").getLines
        emoji.map { l =>
          val a = l.split(' ')
          (a(1), a(0))
        }.toMap
      } getOrElse Map.empty
    }

    private def emoji(line: String): (String, Int) = {
      var drop = 0

      val content: String = emojiEscape.replaceAllIn(line, m => {
        m.group(1) match {
          case e if emojis.contains(e) => drop += m.matched.length - 1; emojis(e)
          case _ => m.matched
        }
      })

      (content, drop)
    }

    def apply(line: String): Line = {
      val (l1, lineStyle) = style(line)
      val (l2, ansiDrop) = ansi(l1)
      val (content, emojiDrop) = emoji(l2)

      val length = l1.codePointCount(0, l1.length) - ansiDrop - emojiDrop

      Line(content = content, length = length, style = lineStyle)
    }
  }

  // `size` and `maxLength` refer to the dimensions of the slide's last build
  private case class Build(content: IndexedSeq[Line], size: Int, maxLength: Int, footer: Line)

  private case class Slide(content: IndexedSeq[Line], builds: IndexedSeq[Int], code: IndexedSeq[String]) {
    private val maxLength = content.maxBy(_.length).length

    def lastBuild: Int = builds.size - 1
    def hasBuild(n: Int): Boolean = builds.isDefinedAt(n)
    def build(n: Int, footer: Line): Build = Build(content.take(builds(n)), content.size, maxLength, footer)
  }

  private case class Deck(slides: IndexedSeq[Slide]) {
    private var slideCursor = -1
    private var buildCursor = 0

    private def currentSlideIsDefined: Boolean = slides.isDefinedAt(slideCursor)
    private def currentSlide: Slide = slides(slideCursor)

    private def footer: Line = {
      val sb = StringBuilder.newBuilder
      if (slideCounter) {
        sb ++= ">> " + (slideCursor + 1)

        if (slideTotal) sb ++= "/" + slides.size

        sb ++= " "
      }

      Line(sb.mkString)
    }

    private def select(slide: Int = slideCursor, build: Int = 0): Option[Build] = {
      // "Stops" the cursor one position after/before the last/first slide to avoid
      // multiple next/previous calls taking it indefinitely away from the deck
      slideCursor = slide.min(slides.size).max(-1)

      buildCursor = build

      if (currentSlideIsDefined && currentSlide.hasBuild(buildCursor)) {
        Some(currentSlide.build(buildCursor, footer))
      } else None
    }

    def jumpTo(n: Int): Option[Build] = select(slide = n)

    def jump(n: Int): Option[Build] = jumpTo(slideCursor + n)

    def nextBuild: Option[Build] = select(build = buildCursor + 1) orElse jump(1)

    def redrawBuild: Option[Build] = select(build = buildCursor)

    def previousBuild: Option[Build] = select(build = buildCursor - 1) orElse {
      jump(-1) flatMap { _ =>
        select(build = currentSlide.lastBuild)
      }
    }

    def lastSlide: Option[Build] = jumpTo(slides.size - 1)

    def lastBuild: Option[Build] = jumpTo(slides.size) orElse previousBuild

    def currentSlideNumber: Int = slideCursor

    def runCode: Unit = {
      val code = currentSlide.code(buildCursor)

      if (repl.isEmpty) {
        Console.err.print(s"No reference to REPL found. Please call with parameter intp=$$intp")
      } else if (code.isEmpty) {
        Console.err.print("No code for you")
      } else {
        repl foreach (_.interpret(code))
      }
    }
  }

  private val helpMessage = """Usage:
    |  next          n      >     go to next build/slide
    |  previous      p      <     go back to previous build/slide
    |  redraw        z            redraw the current build/slide
    |  Next          N      >>    go to next slide
    |  Previous      P      <<    go back to previous slide
    |  i next        i n          advance i slides
    |  i previous    i p          go back i slides
    |  i go          i g          go to slide i
    |  first         f      |<    go to first slide
    |  last          l      >|    go to last slide
    |  Last          L      >>|   go to last build of last slide
    |  run           r      !!    execute code that appears on slide
    |  blank         b            blank screen
    |  help          h      ?     print this help message""".stripMargin

  private val repl = Option(intp)

  private var deck = Deck(parseSource(source))

  private def parseSource(path: String): IndexedSeq[Slide] = {
    Try {
      val pathFile = new File(path)
      val lines: Iterator[String] = (
        if (pathFile.isDirectory) {
          pathFile
            .list
            .sorted
            .filter(_.endsWith(".replesent"))
            .flatMap { name => io.Source.fromFile(new File(pathFile, name)).getLines }
            .toIterator
        } else {
          io.Source.fromFile(path).getLines
        }
      )
      parse(lines)
    } match {
      case Failure(e) =>
        e.printStackTrace
        Console.err.print(s"Sorry, could not parse $path. Quick, say something funny before anyone notices!")
        IndexedSeq.empty
      case Success(value) => value
    }
  }

  private def parse(lines: Iterator[String]): IndexedSeq[Slide] = {
    sealed trait LineHandler {
      def switch: LineHandler
      def apply(line: String): (Line, Option[String])
    }

    object LineHandler extends LineHandler {
      def switch: LineHandler = CodeHandler
      def apply(line: String): (Line, Option[String]) = (Line(line), None)
    }

    object CodeHandler extends LineHandler {
      private val patterns: Seq[(String, Regex)] = {
        val number: Regex = {
          val hex = "(?:0[xX][0-9A-Fa-f]+)"
          val decimal = "(?:[1-9][0-9]*|0)"
          val long = s"(?:${decimal}[DFLdfl])"
          val float = s"(?:${decimal}\\.${decimal}[DFdf])"
          val eNotation = s"(?:${decimal}(?:\\.0?${decimal})?[eE][+\\-]?[0-9]+)"
          s"""\\b(?:${eNotation}|${hex}|${long}|${float}|${decimal})\\b""".r
        }

        val string: Regex = "(?:s?\"(?:\\\\\"|[^\"])*\")".r
        val reserved: Regex = (
          s"""\\b(?:null|contains|exists|filter|filterNot|find|flatMap|""" +
          s"""flatten|fold|forall|foreach|getOrElse|map|orElse)\\b"""
        ).r
        val special: Regex = s"""\\b(?:true|false|this)\\b""".r
        val typeSig: Regex = {
          val token: String => String = { limit => s"[$$_]${limit}[A-Z][_$$A-Z0-9]${limit}[\\w$$]${limit}" }
          val prefix: String = s"""(?<=(?::)\\s{0,10}|\\btype ${token("{0,10}")}\\s{0,10}=\\s{0,10})"""
          s"""\\b(?:${prefix}(?:${token("*")}|\\s*=>\\s*|\\s*with\\s*)*)\\b"""
        }.r

        val syntax: Regex = (
          s"""\\b(?:abstract|case|catch|class|def|do|else|extends|final|""" +
          s"""finally|for|forSome|if|implicit|import|lazy|match|new|""" +
          s"""object|override|package|private|protected|return|sealed|""" +
          s"""super|throw|trait|try|type|val|var|while|with|yield)\\b"""
        ).r

        Seq[(String, Regex)](
          "r" -> string
        , "c" -> reserved
        , "m" -> special
        , "g" -> typeSig
        , "r" -> number
        , "y" -> syntax
        )
      }

      def switch: LineHandler = LineHandler

      def apply(line: String): (Line, Option[String]) = {
        val (colors, regexes) = patterns.unzip

        // new Regex("(?:(a)|(b)|(c))") will produce
        // m.subgroups List(null, "b", null) when applied on "b"
        val regex = new Regex(s"(?:(${regexes.mkString(")|(")}))")

        val formatted = regex.replaceAllIn(line, { m =>
          val colorIdx = m.subgroups.indexWhere(_ != null)
          colors.drop(colorIdx).take(1).headOption
            .map({ color =>
              s"\\\\${color}${Regex.quoteReplacement(m.toString)}\\\\s"
            })
            .getOrElse(line)
        })

        (Line("< " + formatted), Option(line))
      }
    }

    case class Acc(
      content: IndexedSeq[Line] = IndexedSeq.empty
    , builds: IndexedSeq[Int] = IndexedSeq.empty
    , deck: IndexedSeq[Slide] = IndexedSeq.empty
    , code: IndexedSeq[String] = IndexedSeq.empty
    , codeAcc: IndexedSeq[String] = IndexedSeq.empty
    , handler: LineHandler = LineHandler
    ) {
      import config.newline

      def switchHandler: Acc = copy(handler = handler.switch)

      def append(line: String): Acc = {
        val (l, c) = handler(line)
        copy(content = content :+ l, codeAcc = c.fold(codeAcc)(codeAcc :+ _))
      }

      def pushBuild: Acc = copy(
        builds = builds :+ content.size
      , code = code :+ codeAcc.mkString(newline)
      , codeAcc = IndexedSeq.empty
      )

      def pushSlide: Acc = {
        if (content.isEmpty) {
          append("").pushSlide
        } else {
          val finalBuild = pushBuild
          val slide = Slide(content, finalBuild.builds, finalBuild.code)

          Acc(deck = deck :+ slide)
        }
      }
    }

    val slideSeparator = "---"
    val buildSeparator = "--"
    val codeDelimiter = "```"

    val acc = lines.foldLeft(Acc()) { (acc, line) =>
      line match {
        case `slideSeparator` => acc.pushSlide
        case `buildSeparator` => acc.pushBuild
        case `codeDelimiter` => acc.switchHandler
        case _ => acc.append(line)
      }
    }.pushSlide

    acc.deck
  }

  private def render(build: Build): String = {
    import config._

    val topPadding = (verticalSpace - build.size) / 2
    val bottomPadding = verticalSpace - topPadding - build.content.size

    val margin = horizontalSpace - build.maxLength

    val sb = StringBuilder.newBuilder

    def render(line: Line): StringBuilder = {
      sb ++= sinistral
      sb ++= line.render(margin)
      sb ++= dextral
      sb ++= newline
    }

    sb ++= topRow
    sb ++= blankLine * topPadding

    build.content foreach render

    if (slideCounter && bottomPadding > 0) {
      sb ++= blankLine * (bottomPadding - 1)
      render(build.footer)
    } else {
      sb ++= blankLine * bottomPadding
    }

    sb ++= bottomRow

    sb.mkString
  }

  private def show(build: Option[Build]): Unit = {
    if (build.isEmpty) Console.err.print("No slide for you")

    build foreach { b =>
      print(render(b))
    }
  }
  private def reloadDeck(): Unit = {
    val curSlide = deck.currentSlideNumber
    deck = Deck(parseSource(source))
    show(deck.jumpTo(curSlide))
  }

  implicit class Ops(val i: Int) {
    def next: Unit = show(deck.jump(i))
    def n: Unit = next

    def previous: Unit = show(deck.jump(-i))
    def p: Unit = previous

    def go: Unit = show(deck.jumpTo(i - 1))
    def g: Unit = go
  }

  def next: Unit = show(deck.nextBuild)
  def n: Unit = next
  def > : Unit = next

  def previous: Unit = show(deck.previousBuild)
  def p: Unit = previous
  def < : Unit = previous

  def redraw: Unit = show(deck.redrawBuild)
  def z: Unit = redraw

  def reload: Unit = reloadDeck
  def y: Unit = reload

  def Next: Unit = 1.next
  def N: Unit = Next
  def >> : Unit = Next

  def Previous: Unit = 1.previous
  def P: Unit = Previous
  def << : Unit = Previous

  def first: Unit = 1.go
  def f: Unit = first
  def |< : Unit = first

  def last: Unit = show(deck.lastSlide)
  def l: Unit = last
  def >| : Unit = last

  def Last: Unit = show(deck.lastBuild)
  def L: Unit = Last
  def >>| : Unit = Last

  def run: Unit = deck.runCode
  def r: Unit = run
  def !! : Unit = run

  def blank: Unit = print(config.newline * config.screenHeight)
  def b: Unit = blank

  def help: Unit = print(helpMessage)
  def h: Unit = help
  def ? : Unit = help
}
